server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,236 @@
+
+        // the class key is to ensure we only load one copy of each implementation
+        // if we're registering the same things multiple times for tests etc
+        var existingSpecs = new HashMap<>(FEATURE_SPECS);
+        for (FeatureSpecification spec : specs) {
+            existingSpecs.put(spec.getClass(), spec);
+        }
+        if (Assertions.ENABLED) {
+            checkFeatureSpecs(existingSpecs.values());
+        }
+        FEATURE_SPECS = Collections.unmodifiableMap(existingSpecs);
+    }
+
+    private static void checkFeatureSpecs(Collection<FeatureSpecification> specs) {
+        Map<String, FeatureSpecification> allFeatures = new HashMap<>();
+        for (FeatureSpecification spec : specs) {
+            for (NodeFeature f : spec.getFeatures()) {
+                var existing = allFeatures.putIfAbsent(f.id(), spec);
+                if (existing != null) {
+                    throw new IllegalArgumentException(
+                        Strings.format("Duplicate feature - [%s] is declared by both [%s] and [%s]", f.id(), existing, spec)
+                    );
+                }
+            }
+
+            for (var hfe : spec.getHistoricalFeatures().entrySet()) {
+                var existing = allFeatures.putIfAbsent(hfe.getKey().id(), spec);
+                if (existing != null) {
+                    throw new IllegalArgumentException(
+                        Strings.format("Duplicate feature - [%s] is declared by both [%s] and [%s]", hfe.getKey().id(), existing, spec)
+                    );
+                }
+
+                if (hfe.getValue().onOrAfter(MAX_HISTORICAL_VERSION_EXCLUSIVE)) {
+                    throw new IllegalArgumentException(
+                        Strings.format(
+                            "Historical feature [%s] declared by [%s] for version [%s] is not a historical version",
+                            hfe.getKey().id(),
+                            spec,
+                            hfe.getValue()
+                        )
+                    );
+                }
+                if (hfe.getKey().era().era() != hfe.getValue().major) {
+                    throw new IllegalArgumentException(
+                        Strings.format(
+                            "Historical feature [%s] declared by [%s] has incorrect feature era [%s] for version [%s]",
+                            hfe.getKey().id(),
+                            spec,
+                            hfe.getKey().era(),
+                            hfe.getValue(),
+                            spec
+                        )
+                    );
+                }
+            }
+        }
+    }
+
+    private static NavigableMap<Version, Set<String>> calculateHistoricalFeaturesFromSpecs() {
+        readSpecs();
+
+        // use sorted sets here so they're easier to look through in info & debug output
+        NavigableMap<Version, NavigableSet<String>> declaredHistoricalFeatures = new TreeMap<>();
```

Comment:
Using a sorted structure means slower access at runtime. If the concern is info/debug output, could we instead sort the data when logging, since that only happens once, and use a hash based structure in memory?

Reply:
Good idea - changed